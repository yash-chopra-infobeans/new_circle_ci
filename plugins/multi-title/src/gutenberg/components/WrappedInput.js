import classNames from 'classnames';
import TextareaAutosize from 'react-autosize-textarea';
import withForwardedRef from '../utils/forwardRef';
import { CharLimit } from './CharLimit';

const { __ } = wp.i18n;
const { Component } = wp.element;

export class WrappedInput extends Component {
  constructor(...props) {
    super(...props);

    this.state = {
      charLimitDisplay: true,
      errors: false,
      errorMessages: [],
    };
  }

  componentDidMount() {
    const { showCharLimitOnFocus } = this.props;

    // hide character limit bar if showCharLimitOnFocus is passed as prop
    if (showCharLimitOnFocus) {
      this.setState({
        charLimitDisplay: false,
      });
    }
  }

  onChange = value => {
    const { name, titleType = 'additional', onChange, charLimitValue = '' } = this.props;

    /** Check input value meets criteria passing in any other characters (charLimitValue)
     * that should be taken into account when validating the input value.
     */
    const validation = this.validateInput(`${charLimitValue}${value}`);

    this.setState({
      values: {
        [titleType]: {
          [name]: onChange(value, validation) || value,
        },
      },
    });
  };

  decodeHTML = value => {
    const text = document.createElement('textarea');
    text.innerHTML = value;

    return text.value;
  };

  validateInput(value) {
    const { charLimit = false, onError } = this.props;
    let errors = false;
    let errorMessages = [];

    // if character limit and value length is greater set appropriate error message
    if (charLimit && value.length > charLimit) {
      errors = true;
      errorMessages = [
        ...errorMessages,
        __('You have exceeded the maximum character limit.', 'multi-title'),
      ];
    }

    // if errors and onError function has been passed call it passing in error state and messages
    if (errors && onError) {
      onError(errors, errorMessages);
    }

    return {
      errors,
      errorMessages,
    };
  }

  render() {
    const {
      name,
      titleType = 'additional',
      charLimit = false,
      charLimitValue = '',
      className = '',
      hideErrors = false,
      onBlur,
      onFocus,
      showCharLimitAt = -1,
      showCharLimitOnFocus,
      value = '',
      rows = 1,
      forwardedRef,
      ...inputProperties
    } = this.props;

    const { values: stateValue = [], charLimitDisplay, errors, errorMessages } = this.state;

    const charLimitClasses = classNames(className, 'wrappedInput', {
      'charLimit--visible':
        charLimitDisplay && charLimit && `${charLimitValue}${value}`.length >= showCharLimitAt,
    });

    let inputValue = value;

    if (stateValue?.[titleType]?.[name]) {
      inputValue = stateValue[titleType][name];
    }

    inputValue = this.decodeHTML(inputValue);

    return (
      <div className={charLimitClasses}>
        <CharLimit
          charLimit={charLimit}
          charLimitValue={`${charLimitValue}${value}`}
          showCharLimitAt={showCharLimitAt}
        />
        <div className="wrappedInput-input">
          <TextareaAutosize
            {...inputProperties}
            name={`${name}-${titleType}`}
            ref={forwardedRef}
            rows={rows}
            value={inputValue}
            onChange={e => this.onChange(e.target.value)}
            onBlur={() => {
              this.setState({
                values: {},
              });

              if (showCharLimitOnFocus) {
                this.setState({
                  charLimitDisplay: false,
                });
              }

              if (onBlur) {
                onBlur(value);
              }
            }}
            onFocus={() => {
              if (showCharLimitOnFocus) {
                this.setState({
                  charLimitDisplay: true,
                });
              }

              if (onFocus) {
                onFocus(value);
              }
            }}
          />
        </div>
        {errors && !hideErrors && (
          <div className="wrappedInput-errors">
            <ul>
              {errorMessages.map(message => (
                <li>{message}</li>
              ))}
            </ul>
          </div>
        )}
      </div>
    );
  }
}

export default withForwardedRef(WrappedInput);
