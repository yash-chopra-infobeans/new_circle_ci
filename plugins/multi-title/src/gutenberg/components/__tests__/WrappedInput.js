import { shallow, mount } from 'enzyme';
import { WrappedInput } from '../WrappedInput';

describe('WrappedInput class', () => {
  it('should render without errors', () => {
    const componentProps = {
    };
    const component = shallow(<WrappedInput {...componentProps} />);

    expect(component).toMatchSnapshot();
  });

  it('should render error in UI', () => {
    const componentProps = {};
    const component = shallow(<WrappedInput {...componentProps} />);

    component.setState({
      errors: true,
      errorMessages: ['You have exceeded the maximum character limit.'],
    });

    const error = component.find('.wrappedInput-errors ul li');

    expect(component).toMatchSnapshot();
    expect(error.text()).toEqual('You have exceeded the maximum character limit.');
  });

  it('should set any additional props as input attribute(s)', () => {
    const componentProps = {
      additionalprop: 'some data',
    };
    const component = mount(<WrappedInput {...componentProps} />);
    const input = component.find('textarea').at(0);

    expect(input.prop('additionalprop')).toEqual('some data');
  });

  describe('input validation', () => {
    it('should clear any errors in state on succesfully validating input', () => {
      const componentProps = {
        charLimit: 5,
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      component.setState({
        errors: true,
        errorMessages: ['You have exceeded the maximum character limit.'],
      });

      expect(component.instance().validateInput('test')).toEqual({
        errors: false,
        errorMessages: [],
      });
    });

    it('should set error state when character limit exceeded', () => {
      const componentProps = {
        charLimit: 5,
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      expect(component.instance().validateInput('morethan5chars')).toEqual({
        errors: true,
        errorMessages: ['You have exceeded the maximum character limit.'],
      });
    });

    it('should call onError when passed as prop and character limit exceeded', () => {
      const componentProps = {
        charLimit: 5,
        onError: jest.fn(),
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      component.instance().validateInput('morethan5chars');
      expect(componentProps.onError).toHaveBeenCalledTimes(1);
    });
  });

  describe('input interactions', () => {
    it('should update state when value changes', () => {
      const componentProps = {
        onChange: jest.fn(),
        name: 'test',
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      component.instance().onChange('more');
      expect(component.state().values).toEqual({
        additional: {
          test: "more",
        }
      });
    });

    it('should update state with returned value from onChange prop when value changes', () => {
      const componentProps = {
        onChange: () => 'test value',
        name: 'test',
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      component.instance().onChange('more');
      expect(component.state().values).toEqual({
        additional: {
          test: "test value",
        }
      });
    });

    it('should call onChange prop when value changes', () => {
      const componentProps = {
        onChange: jest.fn(),
      };
      const component = shallow(<WrappedInput {...componentProps} />);

      component.instance().onChange('more');
      expect(componentProps.onChange).toHaveBeenCalledTimes(1);
    });

    it('should call onFocus when input is in focus', () => {
      const componentProps = {
        onFocus: jest.fn(),
      };
      const component = mount(<WrappedInput {...componentProps} />);
      const input = component.find('textarea').at(0);

      input.simulate('focus');

      expect(componentProps.onFocus).toHaveBeenCalledTimes(1);
    });

    it('should call onBlur when input has lossed focus', () => {
      const componentProps = {
        onBlur: jest.fn(),
        onChange: jest.fn(),
      };
      const component = mount(<WrappedInput {...componentProps} />);
      const input = component.find('textarea').at(0);

      input.simulate('focus');
      input.simulate('change');
      input.simulate('blur');

      expect(componentProps.onBlur).toHaveBeenCalledTimes(1);
    });

    it('should display character limit bar on input focus', () => {
      const componentProps = {
        onBlur: jest.fn(),
        onChange: jest.fn(),
        showCharLimitOnFocus: true,
        value: 'ss',
      };
      const component = mount(<WrappedInput {...componentProps} />);
      const input = component.find('textarea').at(0);

      expect(component.state().charLimitDisplay).toEqual(false);

      input.simulate('focus');

      expect(component.state().charLimitDisplay).toEqual(true);

      input.simulate('change');
      input.simulate('blur');

      expect(component.state().charLimitDisplay).toEqual(false);
    });
  });
});
