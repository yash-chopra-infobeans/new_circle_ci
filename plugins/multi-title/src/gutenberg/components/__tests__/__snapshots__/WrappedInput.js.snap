// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WrappedInput class should render error in UI 1`] = `
<div
  className="wrappedInput"
>
  <CharLimit
    charLimit={false}
    charLimitValue=""
    showCharLimitAt={-1}
  />
  <div
    className="wrappedInput-input"
  >
    <ForwardRef
      name="undefined-additional"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      rows={1}
      value=""
    />
  </div>
  <div
    className="wrappedInput-errors"
  >
    <ul>
      <li>
        You have exceeded the maximum character limit.
      </li>
    </ul>
  </div>
</div>
`;

exports[`WrappedInput class should render without errors 1`] = `
<div
  className="wrappedInput"
>
  <CharLimit
    charLimit={false}
    charLimitValue=""
    showCharLimitAt={-1}
  />
  <div
    className="wrappedInput-input"
  >
    <ForwardRef
      name="undefined-additional"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      rows={1}
      value=""
    />
  </div>
</div>
`;
